<h1><center> <font color='brown'> BDAT 1004 – Data Programming</font></center></h1>
<h2><center> Problem Set 2</center></h2>
<h2><center> Prepared by Arsha Unnithan (200546119)</center></h2>

<h3><font color='darkblue'>Question 1 </h3>

<b>What value is displayed when the last expression (a) is evaluated? Explain your answer by indicating what happens in every executed statement.</b>

In [1]:
a=0

def b():
    global a 
    a = c(a)
    
def c(a):
    return a + 2

In [2]:
b()
b()
b()
a

6

Explanation:

This code defines three functions b(), c(a), and a global variable a.

Initially, the value of a is set to 0.

The b() function uses the global keyword to indicate that the a variable being referred to inside the function is the global one. The function calls the c(a) function, passing the value of a to it. c(a) function then returns the value of a + 2, which is then assigned to the global variable a.

The b() function is then called three times in a row. Each time it is called, it updates the value of a by calling the c(a) function and adding 2 to the current value of a. After being called three times, a has the value of 6.

Finally, the code outputs the value of a using the print() function, which is 6.

.

<h3><font color='darkblue'>Question 2 </h3>

<b>Function fileLength(), given to you, takes the name of a file as input and returns the length of the file.</b>

In [17]:
def filelength(file):
    try:
        file=open(file)
        contents=file.read()
        file.close()
        print(len(contents))
    except:
        FileNotFoundError()
        print(f'File "{file}" not found.')

In [16]:
filelength('FileLength.txt')

123


In [15]:
filelength('findit.txt')

File "findit.txt" not found.


.

<h3><font color='darkblue'>Question 3 </h3>

<b>Write a class named Marsupial that can be used as shown. Now write a class named Kangaroo as a subclass of Marsupial that inherits all the attributes of Marsupial.</b>

In [22]:
class Marsupial:
    def __init__(self):
        self.myLst = []
        
    def put_in_pouch(self,item):       
        self.myLst.append(item)
        
    def pouch_contents(self):          
        return self.myLst
    
m = Marsupial()
m.put_in_pouch('doll')
m.put_in_pouch('firetruck')
m.put_in_pouch('kitten')
m.pouch_contents()

['doll', 'firetruck', 'kitten']

In [23]:
class Kangaroo(Marsupial):
    def __init__(self,x,y):        
        Marsupial.__init__(self)
        self.x = x
        self.y = y
        
    def jump(self,dx,dy):          
        self.x += dx
        self.y += dy
        
    def __str__(self):            
        return 'I am a Kangaroo located at co-ordinates ({},{})'.format(self.x,self.y)

k = Kangaroo(0,0)
print(k)

I am a Kangaroo located at co-ordinates (0,0)


In [24]:
k.put_in_pouch('doll')
k.put_in_pouch('firetruck')
k.put_in_pouch('kitten')
km=k.pouch_contents()
print(km)

['doll', 'firetruck', 'kitten']


In [25]:
k.jump(1,0)
k.jump(1,0)
k.jump(1,0)

print(k)

I am a Kangaroo located at co-ordinates (3,0)


.

<h3><font color='darkblue'>Question 4 </h3>

<b>Write function collatz() that takes a positive integer x as input and prints the Collatz sequence starting at x. <b>Write a class named Marsupial that can be usibutes of Marsupial.</b>

In [26]:
def collatz(x):
    if x == 1:
        print(x)
    elif x % 2 == 0:
        print(x)
        collatz(x//2)
    else:
        print(x)
        collatz(3*x + 1)

In [27]:
collatz(1)

1


In [29]:
collatz(10) 

10
5
16
8
4
2
1


In [34]:
collatz(8) 

8
4
2
1


.

<h3><font color='darkblue'>Question 5</h3>

<b>Write a recursive method binary() that takes a non-negative integer n and prints the binary representation of integer n.

In [44]:
def binary(n):         
    if n > 1:
        binary(n // 2)
        
    print(n % 2, end='')

In [45]:
binary(0) 

0

In [46]:
 binary(1) 

1

In [47]:
binary(3)

11

In [48]:
binary(9)

1001

.

<h3><font color='darkblue'>Question 6</h3>

<b> Implement a class named HeadingParser that can be used to parse an HTML document, and retrieve and print all the headings in the document. You should implement your class as a subclass of HTMLParser, defined in Standard Library module html.parser.

In [59]:
from html.parser import HTMLParser 

class HeadingParser(HTMLParser):
    indent = 0
    headerFlag = False
    
    
    def handle_starttag(self, tag, attrs):  
        if tag == "h1":
            self.headerFlag = True
            self.indent= 0
        elif tag == "h2":
            self.headerFlag = True
            self.indent= 1
        else:
            self.headerFlag = False
            self.indent= 0
            
    def handle_data(self, w3c_data):                
        if self.headerFlag: 
            to_indent = " "*self.indent
            print('{} {}'.format(to_indent,w3c_data))
            
    def handle_endtag(self, tag):
        if tag in ["h1","h2"]:
            self.headerFlag = False

In [60]:
infile = open('w3c.html")
content = infile.read()
infile.close()
hp = HeadingParser()
hp.feed(content)

 W3C Mission
  Principles


.

<h3><font color='darkblue'>Question 7</h3>

<b> Implement recursive function webdir() that takes as input: a URL (as a string) and non-negative integers depth and indent. Your function should visit every web page reachable from the starting URL web page in depth clicks or less, and print each web page's URL. As shown below, indentation, specified by indent,  should be used to indicate the depth of a URL.

In [100]:
from html.parser import HTMLParser    
from urllib.request import urlopen
from urllib.parse import urljoin

class Parser(HTMLParser):

    def __init__(self, url):
        HTMLParser.__init__(self)
        self.url = url
        self.url_list = []

    def append_list(self):
        return self.url_list

    def handle_starttag(self, tags, attri):
        for head in attri:
            contents = urljoin(self.url, head[1])
            if contents[:4] == 'http': 
                self.url_list.append(contents)
indentation = 0
def webdir(url, distance, indentation):
   

    distance -= 1    
    print(indentation*'  ', url)         

    object = urlopen(url).read().decode()
    column = Parser(url)
    column.feed(object)
    url_link = column.append_list() 
 
    url_list = url_link
    indentation += 4

    for links in url_list:
        if distance < 0 or indentation < 0:
            return 1
        else:
            webdir(links, distance, indentation)            

In [107]:
webdir('http://reed.cs.depaul.edu/lperkovic/csc242/test1.html' , 2, 0) 
# the link provided in the assignemnt is not giving the output as expected due to HTTPError

 http://reed.cs.depaul.edu/lperkovic/csc242/test1.html


HTTPError: HTTP Error 404: 

HTTP Status 404 – Not Found
Type Status Report

Message The requested resource [/lperkovic/csc242/test1.html] is not available

Description The origin server did not find a current representation for the target resource or is not willing to disclose that one exists.

Apache Tomcat/9.0.50

.

<h3><font color='darkblue'>Question 8</h3>

<b> Write SQL queries on the below database table that return: 

In [108]:
import sqlite3
con = sqlite3.connect('WeatherData.db')
cur = con.cursor()

cur.execute("CREATE TABLE WEATHER_DATA (City text, Country text, Season text, Temperature float, Rainfall float)")
data=[('Mumbai', 'India','Winter',24.8, 5.9),('Mumbai','India','Spring',28.4,16.2),('Mumbai','India','Summer',27.9,1549.4),
     ('Mumbai','India','Fall',27.6,346.0),('London','United Kingdom','Winter',4.2,207.7),
     ('London','United Kingdom','Spring',8.3,169.6),('London','United Kingdom','Summer',15.7,157.0),
     ('London','United Kingdom','Fall',10.4,218.5),('Cairo','Egypt','Winter',13.6,16.5),('Cairo','Egypt','Spring',20.7,6.5),
     ('Cairo','Egypt','Summer',27.7,0.1),('Cairo','Egypt','Fall',22.2,4.5)]

for cities in data:
    cur.execute("INSERT INTO WEATHER_DATA VALUES (?,?,?,?,?)",cities)
    con.commit()

<b>a) All the temperature data.             

In [111]:
cur.execute('SELECT Temperature FROM WEATHER_DATA')
print(cur.fetchall())

[(24.8,), (28.4,), (27.9,), (27.6,), (4.2,), (8.3,), (15.7,), (10.4,), (13.6,), (20.7,), (27.7,), (22.2,)]


<b>b) All the cities, but without repetition.

In [112]:
cur.execute('SELECT DISTINCT City FROM WEATHER_DATA')
print(cur.fetchall())

[('Mumbai',), ('London',), ('Cairo',)]


<b>c) All the records for India.

In [126]:
cur.execute('SELECT * FROM WEATHER_DATA WHERE Country LIKE "India%"')
cur.fetchall()

[('Mumbai', 'India', 'Winter', 24.8, 5.9),
 ('Mumbai', 'India', 'Spring', 28.4, 16.2),
 ('Mumbai', 'India', 'Summer', 27.9, 1549.4),
 ('Mumbai', 'India', 'Fall', 27.6, 346.0)]

<b>d) All the Fall records.

In [116]:
cur.execute('SELECT * FROM WEATHER_DATA WHERE Season LIKE "Fall%"')
cur.fetchall()

[('Mumbai', 'India', 'Fall', 27.6, 346.0),
 ('London', 'United Kingdom', 'Fall', 10.4, 218.5),
 ('Cairo', 'Egypt', 'Fall', 22.2, 4.5)]

<b>e) The city, country, and season for which the average rainfall is between 200 and 400 millimeters.

In [119]:
cur.execute('SELECT City, Country, Season FROM WEATHER_DATA WHERE Rainfall BETWEEN 200 AND 400')
cur.fetchall()

[('Mumbai', 'India', 'Fall'),
 ('London', 'United Kingdom', 'Winter'),
 ('London', 'United Kingdom', 'Fall')]

<b>f) The city and country for which the average Fall temperature is above 20 degrees, in increasing temperature order. 

In [123]:
cur.execute('SELECT City, Country,Temperature FROM WEATHER_DATA WHERE Temperature > 20 ORDER BY Temperature DESC')
cur.fetchall()

[('Mumbai', 'India', 28.4),
 ('Mumbai', 'India', 27.9),
 ('Cairo', 'Egypt', 27.7),
 ('Mumbai', 'India', 27.6),
 ('Mumbai', 'India', 24.8),
 ('Cairo', 'Egypt', 22.2),
 ('Cairo', 'Egypt', 20.7)]

<b>g) The total annual rainfall for Cairo.

In [127]:
cur.execute('SELECT SUM(Rainfall) AS RF, City FROM WEATHER_DATA WHERE City LIKE "Cairo" GROUP BY City')
cur.fetchall()

[(27.6, 'Cairo')]

<b>h) The total rainfall for each season.

In [131]:
cur.execute('SELECT SUM(Rainfall),Season FROM WEATHER_DATA GROUP BY Season')
cur.fetchall()

[(569.0, 'Fall'),
 (192.29999999999998, 'Spring'),
 (1706.5, 'Summer'),
 (230.1, 'Winter')]

.

<h3><font color='darkblue'>Question 9</h3>

<b>Suppose list words is defined as follows: >>> words = ['The', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog'] Write list comprehension expressions that use list words and generate the following lists: 

In [135]:
words = ['The', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog']

<b>a) ['THE', 'QUICK', 'BROWN', 'FOX', 'JUMPS', 'OVER', 'THE', 'LAZY', 'DOG']

In [140]:
print([word.upper() for word in words])

['THE', 'QUICK', 'BROWN', 'FOX', 'JUMPS', 'OVER', 'THE', 'LAZY', 'DOG']


<b>b) ['the', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog']  

In [142]:
print([word.lower() for word in words])

['the', 'quick', 'brown', 'fox', 'jumps', 'over', 'the', 'lazy', 'dog']


<b>c)the list of lengths of words in list words.

In [145]:
print([len(word) for word in words])

[3, 5, 5, 3, 5, 4, 3, 4, 3]


<b>d)the list containing a list for every word of list words, where each list contains the word in uppercase and lowercase and the length of the word.

In [148]:
print([[word.upper(), word.lower(), len(word)] for word in words])

[['THE', 'the', 3], ['QUICK', 'quick', 5], ['BROWN', 'brown', 5], ['FOX', 'fox', 3], ['JUMPS', 'jumps', 5], ['OVER', 'over', 4], ['THE', 'the', 3], ['LAZY', 'lazy', 4], ['DOG', 'dog', 3]]


<b>e)the list of words in list words containing 4 or more characters.

In [149]:
print([word for word in words if len(word) >= 4])

['quick', 'brown', 'jumps', 'over', 'lazy']


.

--------------------------------------------------------------------------------------------------------------------------------